plugin_tests/sources_test.py

```
@@ -0,0 +1,182 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#############################################################################
+
+import numpy
+import os
+import PIL.Image
+
+from girder import config
+from tests import base
+
+from . import common
+
+
+# boiler plate to start and stop the server
+
+os.environ['GIRDER_PORT'] = os.environ.get('GIRDER_TEST_PORT', '20200')
+config.loadConfig()  # Must reload config to pickup correct port
+
+
+def setUpModule():
+    base.enabledPlugins.append('large_image')
+    base.startServer(False)
+
+
+def tearDownModule():
+    base.stopServer()
+
+
+class LargeImageSourcesTest(common.LargeImageCommonTest):
+    def testMagnification(self):
+        file = self._uploadFile(os.path.join(
+            os.environ['LARGE_IMAGE_DATA'], 'sample_jp2k_33003_TCGA-CV-7242-'
+            '11A-01-TS1.1838afb1-9eee-4a70-9ae3-50e3ab45e242.svs'))
+        itemId = str(file['itemId'])
+        item = self.model('item').load(itemId, user=self.admin)
+        source = self.model('image_item', 'large_image').tileSource(item)
+        mag = source.getMagnification()
+        self.assertEqual(mag['magnification'], 40.0)
+        self.assertEqual(mag['mm_x'], 0.000252)
+        self.assertEqual(mag['mm_y'], 0.000252)
+        mag = source.getMagnificationForLevel()
+        self.assertEqual(mag['magnification'], 40.0)
+        self.assertEqual(mag['mm_x'], 0.000252)
+        self.assertEqual(mag['mm_y'], 0.000252)
+        self.assertEqual(mag['level'], 7)
+        mag = source.getMagnificationForLevel(0)
+        self.assertEqual(mag['magnification'], 0.3125)
+        self.assertEqual(mag['mm_x'], 0.032256)
+        self.assertEqual(mag['mm_y'], 0.032256)
+        self.assertEqual(mag['level'], 0)
+        self.assertEqual(source.getLevelForMagnification(), 7)
+        self.assertEqual(source.getLevelForMagnification(exact=True), None)
```

The above two usages are weird. Maybe `source.getLevelForMagnification(exact=True)` should also return the maximum resolution level.

Alternatively, you can throw an exception if none of magnification, mm_x, and mm_y are given. People can always get the maximum resolution level via `getMetadata`. 

Also is levels 0-based index or a 1-based index?